File: frontend/src/lib/api/market.ts
import { apiClient } from "./client";

export interface MarketIndex {
  symbol: string;
  name: string;
  short: string;
  price: number | null;
  change: number | null;
  change_pct: number | null;
  as_of?: string;
}

export interface MarketMover {
  symbol: string;
  company: string;
  close: string;
  previous_close: string;
  change_pct: string;
  rank: number;
}

export interface MarketMoversResponse {
  date: string | null;
  gainers: MarketMover[];
  losers: MarketMover[];
}

export const marketApi = {
  getIndices: async (): Promise<MarketIndex[]> => {
    const { data } = await apiClient.get<MarketIndex[]>("/market/indices");
    return data;
  },

  getMovers: async (limit = 10): Promise<MarketMoversResponse> => {
    const { data } = await apiClient.get<MarketMoversResponse>("/market/movers", {
      params: { limit },
    });
    return data;
  },
};

File: backend/app/schemas/market.py
"""Pydantic schemas for market-wide endpoints (indices, gainers/losers)."""
from __future__ import annotations

from datetime import date as date_type
from decimal import Decimal

from pydantic import BaseModel


class MarketMoverItem(BaseModel):
    symbol: str
    company: str
    close: Decimal
    previous_close: Decimal
    change_pct: Decimal
    rank: int

    model_config = {"from_attributes": True}


class MarketMoversResponse(BaseModel):
    date: date_type | None
    gainers: list[MarketMoverItem]
    losers: list[MarketMoverItem]
